File: algorithmic_toolbox/week3_greedy_algorithms/4_collecting_signatures/CoveringSegments.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class CoveringSegments {
  private static class Segment {
    int start, end;

    Segment(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  private static int[] optimalPoints(Segment[] segments) {
    List<Integer> points = new ArrayList<>();

    Arrays.sort(segments, (a, b) -> Integer.compare(a.end, b.end));

    while (segments.length > 0) {
      int point = segments[0].end;
      points.add(point);

      int i = 0;
      while (i < segments.length && segments[i].start <= point && segments[i].end >= point)
        i++;

      segments = Arrays.copyOfRange(segments, i, segments.length);
    }

    return points.stream().mapToInt(i -> i).toArray();
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    Segment[] segments = new Segment[n];
    for (int i = 0; i < n; i++) {
      int start, end;
      start = scanner.nextInt();
      end = scanner.nextInt();
      segments[i] = new Segment(start, end);
    }
    scanner.close();

    int[] points = optimalPoints(segments);

    System.out.println(points.length);
    for (int point : points)
      System.out.print(point + " ");
  }
}

File: algorithmic_toolbox/week3_greedy_algorithms/4_collecting_signatures/covering_segments.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
  int start, end;
} Segment;

int compare_segment(const void* a, const void* b) {
  Segment* s1 = (Segment*)a;
  Segment* s2 = (Segment*)b;

  return s1->end - s2->end;
}

int* optimal_points(Segment* segments, int n, int* len) {
  qsort(segments, n, sizeof(Segment), compare_segment);

  int* points = malloc(sizeof(int) * n);
  int point_count = 0;

  while (n > 0) {
    int point = segments[0].end;
    points[point_count++] = point;

    for (int i = 0; i < n; ++i) {
      if (segments[i].start <= point && point <= segments[i].end) {
        segments[i] = segments[--n];
        i--;
      }
    }

    segments = realloc(segments, sizeof(Segment) * n);
  }

  *len = point_count;
  return points;
}

int main() {
  int n;
  int _ = scanf("%d", &n);

  Segment* segments = malloc(sizeof(Segment) * n);
  for (int i = 0; i < n; i++) {
    int _ = scanf("%d %d", &segments[i].start, &segments[i].end);
  }

  int point_count;
  int* points = optimal_points(segments, n, &point_count);
  printf("%d\n", point_count);

  for (int i = 0; i < point_count; i++) {
    printf("%d ", points[i]);
  }
  printf("\n");

  free(segments);
  free(points);

  return 0;
}
